src/Autofac/Core/Lifetime/LifetimeScope.cs

```diff
@@ -150,6 +165,11 @@ private void RaiseBeginning(ILifetimeScope scope)
             handler?.Invoke(this, new LifetimeScopeBeginningEventArgs(scope));
         }
 
+        public void AttachTrace(IResolvePipelineTracer tracer)
+        {
+            _tracer = tracer ?? throw new ArgumentNullException(nameof(tracer));
```

Comment:
Regarding multiple tracers, in my head I saw someone (possibly us), providing a `CompositeTracer` of sorts, that just forwards trace events to a set of contained tracers. Means we don't need to track/maintain a list of them internally.

I put the docs for `AttachTrace` on the interface (and forgot to add `<inheritdoc />` on the implementation, shouldn't that have been caught by a compiler warning?). I put details about the replacement behaviour in remarks, but intellisense doesn't show that, so I've put some words in the summary content. Let me know if that's enough.

Reply:
As for DiagnosticSource, I think we could ship a `DiagnosticSourceTracer` that emitted activity into a provided `DiagnosticSource`, but it might contain less detail than the default tracer, and wouldn't build strings in the same way (because of the perf concerns).

In production, you might want to raise an Activity for each `ResolveRequest`, but we might not track individual middleware executions.

Reply:
I don't think the doc issue will cause a warning because CS1591 (missing docs) is turned off at the project level. It might be worth (later) removing that exclusion to ensure we get full API doc coverage.

I was thinking more about switching the tracing around - we _only_ omit the DiagnosticSource style events, include the information that the tracer would normally need to generate the strings, and if you want to attach a tracer, you register with the .NET diagnostics rather than attaching a tracer directly to anything. Ostensibly we could still supply the default tracer you have and maybe have a convenience method that registers it correctly as a one-liner.

I don't think it's enough to hold this up. And maybe it's not the best idea, I'm not totally sure. I can add an issue for later investigation pre-6.0.